src/librustc_resolve/lib.rs

```diff
@@ -3272,17 +3273,19 @@ impl<'a> Resolver<'a> {
                name == keywords::SelfType.name() && i != 0 ||
                name == keywords::Super.name() && i != 0 ||
                name == keywords::Extern.name() && i != 0 ||
-               name == keywords::Crate.name() && i != 1 &&
-                    path[0].name != keywords::CrateRoot.name() {
+               // we allow crate::foo and ::crate::foo but nothing else
+               name == keywords::Crate.name() && i > 1 &&
+                    path[0].name != keywords::CrateRoot.name() ||
+               name == keywords::Crate.name() && path.len() == 1 {
                 let name_str = if name == keywords::CrateRoot.name() {
                     format!("crate root")
                 } else {
                     format!("`{}`", name)
                 };
                 let msg = if i == 1 && path[0].name == keywords::CrateRoot.name() {
                     format!("global paths cannot start with {}", name_str)
-                } else if i == 0 && name == keywords::Crate.name() {
-                    format!("{} can only be used in absolute paths", name_str)
+                } else if name == keywords::Crate.name() && path.len() == 1 {
+                    "`crate` in paths must be followed by a module name".into()
```

Comment:
`crate` can be used as a single segment path, e.g. in `use crate as something.`
If this line is removed then `keyword-crate-as-identifier.rs` should report a correct error, something like "expected a unit struct or constand, found module".

Reply:
Ah. It was, but I thought that was a bug. Will fix.